### PR TITLE
interfaces/builtin: add interface for remoteproc

### PR DESCRIPTION
**Pull request by bugraaydogar**
This is a new interface that is going to allow developers to interact with Remote Processor Framework of Linux Kernel. Typically, the modern ARM based silicons might have additional Cortex M4 or M7 based microcontroller within the SoC. The `remoteproc` interface will allow a snap to load a firmware to the microcontrollers via snaps. Also, the microcontroller can be started and stopped. Therefore, I implemented this interface as super-privileged. It is up to security review to decide if that makes sense or not.

The documentation related to the sysfs interface is available => https://www.kernel.org/doc/Documentation/ABI/testing/sysfs-class-remoteproc

Thank you,
Bugra

### CODE REVIEW

File: interfaces/builtin/remoteproc.go [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+
+# Allow adding new path for firmware location.
+/sys/module/firmware_class/parameters/path rw,
+`
+
+func init() {
+	registerIface(&commonInterface{
+		name:                  "remoteproc",
+		summary:               remoteprocSummary,
+		implicitOnCore:        true,
```
Comment (zyga) [inline interfaces/builtin/remoteproc.go:55]: Should this be automatic on core or should it be declared in the gadget? In a way only certain devices have remote co-processors but on the other hand we rarely apply this practice to interfaces.

File: interfaces/builtin/remoteproc.go [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+    allow-installation:
+      slot-snap-type:
+        - core
+    deny-auto-connection: true
+`
+
+const remoteprocConnectedPlugAppArmor = `
+# Description: Remote Proc sysfs interface.
+
+/sys/devices/platform/**/remoteproc/remoteproc[0-9]/firmware rw,
```
Comment (zyga) [inline interfaces/builtin/remoteproc.go:41]: Are any of the items matched by `**` symbolic links? Have you verified this on a real device?

File: interfaces/builtin/remoteproc.go [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+/sys/devices/platform/**/remoteproc/remoteproc[0-9]/state rw,
+/sys/devices/platform/**/remoteproc/remoteproc[0-9]/name r,
+/sys/devices/platform/**/remoteproc/remoteproc[0-9]/coredump rw,
+/sys/devices/platform/**/remoteproc/remoteproc[0-9]/recovery rw,
+
+# Allow adding new path for firmware location.
+/sys/module/firmware_class/parameters/path rw,
```
Comment (zyga) [inline interfaces/builtin/remoteproc.go:48]: How is this related to remote processors?

**Review by zyga — COMMENTED** [review]
Quick pass

File: interfaces/builtin/remoteproc.go (hunk 3 above)
Comment (bugraaydogar) [reply interfaces/builtin/remoteproc.go:48]: The default firmware path is `/lib/firmware` and managed by kernel snap. As an application developer, if you want your kernel to look for external firmware is a different path, you would need to pass this path to `/sys/module/firmware_class/parameters/path`. So, this allows developers to load a firmware from the path they have decided.

File: interfaces/builtin/remoteproc.go (hunk 1 above)
Comment (bugraaydogar) [reply interfaces/builtin/remoteproc.go:55]: There is no slot needed to be declared in the gadget snap. The interface will be provided by the snapd.

File: interfaces/builtin/remoteproc.go (hunk 3 above)
Comment (bugraaydogar) [reply interfaces/builtin/remoteproc.go:48]: Correct, this is why I think this interface must be super-privilege.

Comment (pedronis) [reply interfaces/builtin/remoteproc.go:48]: I still do not think that this belongs here, but to a complementary hopefully preexisting interface, as the scope of this - super-privileged or not - is much larger that what the interface might make one think. I'm interested in security feedback on this

Comment (alexmurray) [reply interfaces/builtin/remoteproc.go:48]: I agree - this feels far to broad / non-specific for this current interface interface.

**Review by alexmurray — CHANGES_REQUESTED** [review]
LGTM other than the firmware_class rule - this should be removed as it is not specific to remoteproc.

File: interfaces/builtin/remoteproc.go (hunk 3 above)
Comment (pedronis) [reply interfaces/builtin/remoteproc.go:48]: As discussed previously I think this could go under kernel-module-control or perhaps a new kernel-firmware-control ?

cc @alexmurray

**Comment by bugraaydogar** [comment]
Thank you for the review. I updated the PR to remove `firmware_class` path. I will consider expanding an existing interface with the guidance of @pedronis and @alexmurray.

File: interfaces/builtin/remoteproc.go (hunk 3 above)
Comment (alexmurray) [reply interfaces/builtin/remoteproc.go]: I think a separate interface since I would rather not grant full kernel-module-control when it is not needed.

**Review by zyga — APPROVED** [review]
With the reduced surface, this looks good to me as-is.

@bugraaydogar I understand you will propose the `kernel-firmware-control` interface in separate pull request, correct?

**Review by pedronis — APPROVED** [review]
lgtm

**Review by alexmurray — APPROVED** [review]
LGTM!

**Comment by ernestl** [comment]
@bugraaydogar , thanks for this contribution. Would you please rebase and resolve conflict(s) to enable us to merge?
